--- a/Block/Config.php
+++ b/Block/Config.php
@@ -111,6 +111,10 @@ class Config extends Template
             $config['sandboxSimulationOptions'] = $this->transformSandboxSimulationOptions();
         }
 
+        if ($orderReferenceId = $this->getRequest()->getParam('orderReferenceId')) {
+            $config['orderReferenceId'] = preg_replace('/[^A-Z0-9-]/', '', $orderReferenceId);
+        }
+
         return $config;
     }
 

--- a/Model/AmazonConfig.php
+++ b/Model/AmazonConfig.php
@@ -172,4 +172,15 @@ class AmazonConfig
             $store
         );
     }
+
+    /**
+     * Is AmazonWebapiException code a soft decline error?
+     *
+     * @param $errorCode
+     * @return bool
+     */
+    public function isSoftDecline($errorCode)
+    {
+        return $errorCode == $this->scopeConfig->getValue('payment/amazon_payment/soft_decline_code');
+    }
 }

--- a/composer.json
+++ b/composer.json
@@ -2,7 +2,7 @@
   "name": "amzn/amazon-pay-and-login-with-amazon-core-module",
   "description": "Shared functionality for Amazon Pay and Login with Amazon modules",
   "type": "magento2-module",
-  "version": "3.3.1",
+  "version": "3.4.0",
   "license": [
     "Apache-2.0"
   ],

--- a/etc/module.xml
+++ b/etc/module.xml
@@ -16,7 +16,7 @@
  */
 -->
 <config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Module/etc/module.xsd">
-    <module name="Amazon_Core" setup_version="3.3.1">
+    <module name="Amazon_Core" setup_version="3.4.0">
         <sequence>
             <module name="Magento_Store"/>
             <module name="Magento_Customer"/>

