--- a/Block/Adminhtml/System/Config/Form/SimplepathConfig.php
+++ b/Block/Adminhtml/System/Config/Form/SimplepathConfig.php
@@ -35,6 +35,6 @@ class SimplepathConfig extends \Magento\Config\Block\System\Config\Form\Field
             ->setCacheable(false)
             ->toHtml();
 
-        return $html;
+        return '<div id="row_' . $element->getHtmlId() . '">' . $html . '</div>';
     }
 }

--- a/Block/Adminhtml/System/Config/Form/Version.php
+++ b/Block/Adminhtml/System/Config/Form/Version.php
@@ -56,6 +56,6 @@ class Version extends \Magento\Config\Block\System\Config\Form\Field
         $output = '<div style="background-color:#eee;padding:1em;border:1px solid #ddd;">';
         $output .= __('Module version') . ': ' . $version;
         $output .= "</div>";
-        return $output;
+        return '<div id="row_' . $element->getHtmlId() . '">' . $output . '</div>';
     }
 }

--- a/Domain/AmazonAddressDecoratorJp.php
+++ b/Domain/AmazonAddressDecoratorJp.php
@@ -38,13 +38,7 @@ class AmazonAddressDecoratorJp implements AmazonAddressInterface
      */
     public function getLines()
     {
-        $city = $this->amazonAddress->getCity();
-
-        /*
-         * AmazonAddressDecoratorJp->getCity() returns address line 1 when city is empty.
-         * Omit line 1 from the street address in this case.
-         */
-        return array_slice($this->amazonAddress->getLines(), empty($city) ? 1 : 0);
+        return $this->amazonAddress->getLines();
     }
 
     /**
@@ -76,7 +70,7 @@ class AmazonAddressDecoratorJp implements AmazonAddressInterface
      */
     public function getCity()
     {
-        return $this->amazonAddress->getCity() ?? $this->amazonAddress->getLine(1);
+        return $this->amazonAddress->getCity() ?? '-';
     }
 
     /**

--- a/composer.json
+++ b/composer.json
@@ -2,7 +2,7 @@
   "name": "amzn/amazon-pay-and-login-with-amazon-core-module",
   "description": "Shared functionality for Amazon Pay and Login with Amazon modules",
   "type": "magento2-module",
-  "version": "3.4.1-p3",
+  "version": "3.5.1",
   "license": [
     "Apache-2.0"
   ],

--- a/etc/module.xml
+++ b/etc/module.xml
@@ -16,7 +16,7 @@
  */
 -->
 <config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Module/etc/module.xsd">
-    <module name="Amazon_Core" setup_version="3.4.1-p3">
+    <module name="Amazon_Core" setup_version="3.5.1">
         <sequence>
             <module name="Magento_Store"/>
             <module name="Magento_Customer"/>

--- a/view/adminhtml/templates/system/config/simplepath_admin.phtml
+++ b/view/adminhtml/templates/system/config/simplepath_admin.phtml
@@ -1,4 +1,3 @@
-<div id="amazon_simplepath">
 <!--
 /**
  * Copyright 2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.

