--- a/Helper/Address.php
+++ b/Helper/Address.php
@@ -21,6 +21,7 @@ use Magento\Customer\Api\Data\AddressInterface;
 use Magento\Customer\Api\Data\AddressInterfaceFactory;
 use Magento\Customer\Api\Data\RegionInterfaceFactory;
 use Magento\Directory\Model\RegionFactory;
+use Magento\Framework\App\Config\ScopeConfigInterface;
 
 class Address
 {
@@ -39,14 +40,21 @@ class Address
      */
     private $regionDataFactory;
 
+    /**
+     * @var ScopeConfigInterface
+     */
+    private $scopeConfig;
+
     public function __construct(
         AddressInterfaceFactory $addressFactory,
         RegionFactory $regionFactory,
-        RegionInterfaceFactory $regionDataFactory
+        RegionInterfaceFactory $regionDataFactory,
+        ScopeConfigInterface $config
     ) {
         $this->addressFactory    = $addressFactory;
         $this->regionFactory     = $regionFactory;
         $this->regionDataFactory = $regionDataFactory;
+        $this->scopeConfig = $config;
     }
 
     /**
@@ -58,15 +66,35 @@ class Address
      */
     public function convertToMagentoEntity(AmazonAddressInterface $amazonAddress)
     {
+        $addressLinesAllowed = (int)$this->scopeConfig->getValue(
+            'customer/address/street_lines',
+            \Magento\Store\Model\ScopeInterface::SCOPE_STORE
+        );
+
         $address = $this->addressFactory->create();
         $address->setFirstname($amazonAddress->getFirstName());
         $address->setLastname($amazonAddress->getLastName());
         $address->setCity($amazonAddress->getCity());
-        $address->setStreet(array_values($amazonAddress->getLines()));
         $address->setPostcode($amazonAddress->getPostCode());
         $address->setTelephone($amazonAddress->getTelephone());
         $address->setCountryId($this->getCountryId($amazonAddress));
 
+        /*
+         * The number of lines in a street address is configurable via 'customer/address/street_lines'.
+         * To avoid discarding information, we'll concatenate additional lines so that they fit within the configured
+         *  address length.
+         */
+        $lines = [];
+        for ($i = 1; $i <= 4; $i++) {
+            $line = (string) $amazonAddress->getLine($i);
+            if ($i <= $addressLinesAllowed) {
+                $lines[] = $line;
+            } else {
+                $lines[count($lines)-1] = trim($lines[count($lines)-1] . ' ' . $line);
+            }
+        }
+        $address->setStreet(array_values($lines));
+
         $company = !empty($amazonAddress->getCompany()) ? $amazonAddress->getCompany() : '';
         $address->setCompany($company);
 

--- a/Model/Method/AmazonLoginMethod.php
+++ b/Model/Method/AmazonLoginMethod.php
@@ -51,4 +51,18 @@ class AmazonLoginMethod extends \Magento\Payment\Model\Method\AbstractMethod
      */
     protected $_isOffline = true;
 
+    /**
+     * Is active
+     *
+     * @param int|null $storeId
+     * @return bool
+     */
+    public function isActive($storeId = null)
+    {
+        return (bool)(int)$this->_scopeConfig->getValue(
+            'payment/amazon_pay/active',
+            \Magento\Store\Model\ScopeInterface::SCOPE_STORE,
+            $storeId
+        ) && (bool)(int)$this->getConfigData('active', $storeId);
+    }
 }

--- a/composer.json
+++ b/composer.json
@@ -2,14 +2,14 @@
   "name": "amzn/amazon-pay-module",
   "description": "Amazon Pay module",
   "type": "magento2-module",
-  "version": "2.1.2",
+  "version": "2.1.4",
   "license": [
     "Apache-2.0"
   ],
   "require": {
     "php": "~7.0.13|~7.1.0|~7.2.0",
-    "amzn/amazon-pay-and-login-with-amazon-core-module": "^2.1.2",
-    "amzn/login-with-amazon-module": "^2.1.2",
+    "amzn/amazon-pay-and-login-with-amazon-core-module": "^2.1.4",
+    "amzn/login-with-amazon-module": "^2.1.4",
     "magento/module-eav": "*",
     "magento/module-sales": "*",
     "magento/module-quote": "*",

--- a/etc/module.xml
+++ b/etc/module.xml
@@ -1,6 +1,6 @@
 <?xml version="1.0"?>
 <config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Module/etc/module.xsd">
-    <module name="Amazon_Payment" setup_version="2.1.1" >
+    <module name="Amazon_Payment" setup_version="2.1.4" >
         <sequence>
             <module name="Amazon_Core"/>
             <module name="Amazon_Login"/>

