--- a/Domain/LayoutConfig.php
+++ b/Domain/LayoutConfig.php
@@ -15,10 +15,10 @@
  */
 
 namespace Amazon\Login\Domain;
+
 use Magento\Framework\Model\AbstractModel;
 use Amazon\Core\Helper\Data as CoreHelper;
 
-
 /**
  * Used for accessing Amazon Login layout configuration
  */
@@ -48,5 +48,4 @@ class LayoutConfig
     {
         return !$this->coreHelper->isLwaEnabled();
     }
-
 }

--- a/Test/Unit/Model/LayoutConfigTest.php
+++ b/Test/Unit/Model/LayoutConfigTest.php
@@ -49,5 +49,4 @@ class LayoutConfigTest extends \PHPUnit\Framework\TestCase
             [false, true]
         ];
     }
-
 }

--- a/composer.json
+++ b/composer.json
@@ -2,13 +2,13 @@
   "name": "amzn/login-with-amazon-module",
   "description": "Login with Amazon module",
   "type": "magento2-module",
-  "version": "3.2.13",
+  "version": "3.3.2",
   "license": [
     "Apache-2.0"
   ],
   "require": {
     "php": "~7.1.3||~7.2.0||~7.3.0",
-    "amzn/amazon-pay-and-login-with-amazon-core-module": "^3.2.13",
+    "amzn/amazon-pay-and-login-with-amazon-core-module": "^3.3.2",
     "magento/framework": "^102",
     "magento/module-customer": "^102",
     "magento/module-store": "^101",

--- a/etc/module.xml
+++ b/etc/module.xml
@@ -16,7 +16,7 @@
  */
 -->
 <config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Module/etc/module.xsd">
-    <module name="Amazon_Login" setup_version="3.2.13">
+    <module name="Amazon_Login" setup_version="3.3.2">
         <sequence>
             <module name="Amazon_Core"/>
         </sequence>

--- a/view/frontend/templates/oAuthRedirect.phtml
+++ b/view/frontend/templates/oAuthRedirect.phtml
@@ -18,9 +18,9 @@
 <div>
     <?= $block->escapeHtml(__('You will be redirected shortly...')); ?><br/>
     <?= $block->escapeHtml(__(
-            'If you are not redirected automatically, try refreshing the page,'
+        'If you are not redirected automatically, try refreshing the page,'
             . ' or return to your cart and use another payment method.'
-        )); ?>
+    )); ?>
 </div>
 <script>
 require([

