--- a/composer.json
+++ b/composer.json
@@ -2,14 +2,14 @@
   "name": "amzn/amazon-pay-module",
   "description": "Amazon Pay module",
   "type": "magento2-module",
-  "version": "3.2.7",
+  "version": "3.2.9",
   "license": [
     "Apache-2.0"
   ],
   "require": {
     "php": "~7.1.3||~7.2.0",
-    "amzn/amazon-pay-and-login-with-amazon-core-module": "^3.2.7",
-    "amzn/login-with-amazon-module": "^3.2.7",
+    "amzn/amazon-pay-and-login-with-amazon-core-module": "^3.2.9",
+    "amzn/login-with-amazon-module": "^3.2.9",
     "magento/framework": "^102",
     "magento/module-customer": "^102",
     "magento/module-eav": "^102",

--- a/etc/module.xml
+++ b/etc/module.xml
@@ -16,7 +16,7 @@
  */
 -->
 <config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Module/etc/module.xsd">
-    <module name="Amazon_Payment" setup_version="3.2.7" >
+    <module name="Amazon_Payment" setup_version="3.2.9" >
         <sequence>
             <module name="Amazon_Core"/>
             <module name="Amazon_Login"/>

--- a/view/frontend/web/js/view/shipping-address/inline-form.js
+++ b/view/frontend/web/js/view/shipping-address/inline-form.js
@@ -16,29 +16,25 @@ define([
          */
         initObservable: function () {
             this._super();
-            amazonStorage.isAmazonAccountLoggedIn.subscribe(function (value) {
-                var elem = document.getElementById(this.formSelector);
-
-                if (elem && value === false) {
-                    document.getElementById(this.formSelector).style.display = 'block';
-                }
-            }, this);
-
+            amazonStorage.isAmazonAccountLoggedIn.subscribe(this.hideInlineForm, this);
             return this;
         },
 
         /**
-         * Show/hide inline form
+         * Show/hide inline form depending on Amazon login status
          */
         manipulateInlineForm: function () {
-            var elem;
+            this.hideInlineForm(amazonStorage.isAmazonAccountLoggedIn());
+        },
 
-            if (amazonStorage.isAmazonAccountLoggedIn()) {
-                elem = document.getElementById(this.formSelector);
+        /**
+         * Show/hide inline form
+         */
+        hideInlineForm: function(hide) {
+            var elem = document.getElementById(this.formSelector);
 
-                if (elem) {
-                    document.getElementById(this.formSelector).style.display = 'none';
-                }
+            if (elem) {
+                document.getElementById(this.formSelector).style.display = hide ? 'none' : 'block';
             }
         }
     });

