--- a/composer.json
+++ b/composer.json
@@ -1 +1,37 @@
-{"name":"amzn/amazon-pay-module","description":"Amazon Pay module","type":"magento2-module","version":"3.4.1-p2","license":["Apache-2.0"],"require":{"php":"~7.1.3||~7.2.0||~7.3.0","amzn/amazon-pay-and-login-with-amazon-core-module":"^3.4.1-p2","amzn/login-with-amazon-module":"^3.4.1-p2","magento/framework":"^102","magento/module-customer":"^102","magento/module-eav":"^102","magento/module-sales":"^102","magento/module-quote":"^101.1","magento/module-payment":"^100.3","magento/module-backend":"^101","magento/module-store":"^101","magento/module-checkout":"^100.3","magento/module-catalog":"^103","magento/module-paypal":"*","magento/module-directory":"^100.3"},"suggest":{"magento/module-customer":"*","magento/module-scalable-checkout":"*","magento/module-scalable-oms":"*"},"autoload":{"files":["registration.php"],"psr-4":{"Amazon\\Payment\\":""}}}
+{
+  "name": "amzn/amazon-pay-module",
+  "description": "Amazon Pay module",
+  "type": "magento2-module",
+  "version": "3.4.3",
+  "license": [
+    "Apache-2.0"
+  ],
+  "require": {
+    "php": "~7.1.3||~7.2.0||~7.3.0",
+    "amzn/amazon-pay-and-login-with-amazon-core-module": "^3.4.3",
+    "amzn/login-with-amazon-module": "^3.4.3",
+    "magento/framework": "^102",
+    "magento/module-customer": "^102",
+    "magento/module-eav": "^102",
+    "magento/module-sales": "^102",
+    "magento/module-quote": "^101.1",
+    "magento/module-payment": "^100.3",
+    "magento/module-backend": "^101",
+    "magento/module-store": "^101",
+    "magento/module-checkout": "^100.3",
+    "magento/module-catalog": "^103",
+    "magento/module-paypal": "*",
+    "magento/module-directory": "^100.3"
+  },
+  "suggest": {
+    "magento/module-customer": "*",
+    "magento/module-scalable-checkout": "*",
+    "magento/module-scalable-oms": "*"
+  },
+  "autoload": {
+    "files": ["registration.php"],
+    "psr-4": {
+      "Amazon\\Payment\\": ""
+    }
+  }
+}

--- a/etc/module.xml
+++ b/etc/module.xml
@@ -16,7 +16,7 @@
  */
 -->
 <config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Module/etc/module.xsd">
-    <module name="Amazon_Payment" setup_version="3.4.1" >
+    <module name="Amazon_Payment" setup_version="3.4.3" >
         <sequence>
             <module name="Amazon_Core"/>
             <module name="Amazon_Login"/>

--- a/view/frontend/web/js/model/storage.js
+++ b/view/frontend/web/js/model/storage.js
@@ -50,7 +50,8 @@ define(
             isLoginRedirectPage = $('body').hasClass('amazon-login-login-processauthhash'),
             amazonCustomerEmail = ko.computed(function () {
                 // jscs:disable requireCamelCaseOrUpperCaseIdentifiers
-                if (window.checkoutConfig.hasOwnProperty('amazonLogin') &&
+                if (window.hasOwnProperty('checkoutConfig') &&
+                    window.checkoutConfig.hasOwnProperty('amazonLogin') &&
                     typeof window.checkoutConfig.amazonLogin.amazon_customer_email === 'string'
                 ) {
                     return window.checkoutConfig.amazonLogin.amazon_customer_email;

--- a/view/frontend/web/js/view/shipping.js
+++ b/view/frontend/web/js/view/shipping.js
@@ -2,14 +2,22 @@
 define(
     [
         'jquery',
+        'underscore',
+        'ko',
         'Magento_Checkout/js/view/shipping',
         'Magento_Customer/js/model/customer',
+        'Magento_Checkout/js/action/set-shipping-information',
+        'Magento_Checkout/js/model/step-navigator',
         'Amazon_Payment/js/model/storage'
     ],
     function (
         $,
+        _,
+        ko,
         Component,
         customer,
+        setShippingInformationAction,
+        stepNavigator,
         amazonStorage
     ) {
         'use strict';
@@ -41,20 +49,20 @@ define(
             },
 
             /**
-             * Overridden validateShippingInformation for Amazon Pay to bypass validation
-             *
-             * @inheritDoc
+             * New setShipping Action for Amazon Pay to bypass validation
              */
-            validateShippingInformation: function () {
-                if (!amazonStorage.isAmazonAccountLoggedIn()) {
-                    return this._super();
+            setShippingInformation: function () {
+                if (amazonStorage.isAmazonAccountLoggedIn()) {
+                    if (customer.isLoggedIn() || this.validateGuestEmail()) {
+                        setShippingInformationAction().done(
+                            function () {
+                                stepNavigator.next();
+                            }
+                        );
+                    }
+                } else {
+                    this._super();
                 }
-
-                if (!customer.isLoggedIn()) {
-                    return this.validateGuestEmail();
-                }
-
-                return true;
             }
         });
     }

