--- a/Block/Minicart/Button.php
+++ b/Block/Minicart/Button.php
@@ -30,6 +30,13 @@ use Magento\Framework\App\Request\Http;
  * Class Button
  *
  * @api
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class Button extends Template implements ShortcutInterface
 {

--- a/Block/PaymentLink.php
+++ b/Block/PaymentLink.php
@@ -23,6 +23,13 @@ use Magento\Catalog\Model\Product;
 
 /**
  * @api
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class PaymentLink extends Template
 {

--- a/Block/ProductPagePaymentLink.php
+++ b/Block/ProductPagePaymentLink.php
@@ -25,6 +25,13 @@ use Magento\Framework\View\Element\Template\Context;
 
 /**
  * @api
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class ProductPagePaymentLink extends PaymentLink
 {

--- a/Block/Widget/ResetPassword.php
+++ b/Block/Widget/ResetPassword.php
@@ -17,6 +17,13 @@ use Amazon\Core\Helper\Data;
 
 /**
  * @api
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class ResetPassword extends Template
 {

--- a/Controller/Payment/CompleteCheckout.php
+++ b/Controller/Payment/CompleteCheckout.php
@@ -36,6 +36,13 @@ use Magento\Framework\App\ObjectManager;
  * Class CompleteCheckout
  *
  * @package Amazon\Payment\Controller\Payment
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class CompleteCheckout extends Action
 {

--- a/Controller/Payment/Ipn.php
+++ b/Controller/Payment/Ipn.php
@@ -28,6 +28,14 @@ use Magento\Framework\App\ResponseInterface;
 use Magento\Framework\Exception\NotFoundException;
 use Magento\Framework\App\ObjectManager;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class Ipn extends Action
 {
     /**

--- a/Cron/GetAmazonAuthorizationUpdates.php
+++ b/Cron/GetAmazonAuthorizationUpdates.php
@@ -27,6 +27,14 @@ use Magento\Framework\Api\FilterBuilder;
 use Magento\Framework\Api\Search\FilterGroup;
 use Psr\Log\LoggerInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class GetAmazonAuthorizationUpdates
 {
     /**

--- a/Cron/GetAmazonCaptureUpdates.php
+++ b/Cron/GetAmazonCaptureUpdates.php
@@ -27,6 +27,14 @@ use Magento\Framework\Api\FilterBuilder;
 use Magento\Framework\Api\Search\FilterGroup;
 use Psr\Log\LoggerInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class GetAmazonCaptureUpdates
 {
     /**

--- a/Cron/ProcessAmazonRefunds.php
+++ b/Cron/ProcessAmazonRefunds.php
@@ -22,6 +22,14 @@ use Amazon\Payment\Model\QueuedRefundUpdaterFactory;
 use Amazon\Payment\Model\ResourceModel\PendingRefund\CollectionFactory;
 use Magento\Framework\Data\Collection;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class ProcessAmazonRefunds
 {
     /**

--- a/Domain/AbstractAmazonAuthorizationResponse.php
+++ b/Domain/AbstractAmazonAuthorizationResponse.php
@@ -20,6 +20,14 @@ use Amazon\Payment\Domain\Details\AmazonAuthorizationDetails;
 use Amazon\Payment\Domain\Details\AmazonAuthorizationDetailsFactory;
 use AmazonPay\ResponseInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 abstract class AbstractAmazonAuthorizationResponse
 {
     /**

--- a/Domain/AbstractAmazonCaptureResponse.php
+++ b/Domain/AbstractAmazonCaptureResponse.php
@@ -20,6 +20,14 @@ use Amazon\Payment\Domain\Details\AmazonCaptureDetails;
 use Amazon\Payment\Domain\Details\AmazonCaptureDetailsFactory;
 use AmazonPay\ResponseInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 abstract class AbstractAmazonCaptureResponse
 {
     /**

--- a/Domain/AbstractAmazonRefundResponse.php
+++ b/Domain/AbstractAmazonRefundResponse.php
@@ -20,6 +20,14 @@ use Amazon\Payment\Domain\Details\AmazonRefundDetails;
 use Amazon\Payment\Domain\Details\AmazonRefundDetailsFactory;
 use AmazonPay\ResponseInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 abstract class AbstractAmazonRefundResponse
 {
     /**

--- a/Domain/AbstractAmazonStatus.php
+++ b/Domain/AbstractAmazonStatus.php
@@ -15,6 +15,14 @@
  */
 namespace Amazon\Payment\Domain;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 abstract class AbstractAmazonStatus
 {
     /**

--- a/Domain/AmazonAuthorizationDetailsResponse.php
+++ b/Domain/AmazonAuthorizationDetailsResponse.php
@@ -15,6 +15,14 @@
  */
 namespace Amazon\Payment\Domain;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AmazonAuthorizationDetailsResponse extends AbstractAmazonAuthorizationResponse
 {
     private $resultKey = 'GetAuthorizationDetailsResult';

--- a/Domain/AmazonAuthorizationResponse.php
+++ b/Domain/AmazonAuthorizationResponse.php
@@ -15,6 +15,14 @@
  */
 namespace Amazon\Payment\Domain;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AmazonAuthorizationResponse extends AbstractAmazonAuthorizationResponse
 {
     private $resultKey = 'AuthorizeResult';

--- a/Domain/AmazonAuthorizationStatus.php
+++ b/Domain/AmazonAuthorizationStatus.php
@@ -15,6 +15,14 @@
  */
 namespace Amazon\Payment\Domain;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AmazonAuthorizationStatus extends AbstractAmazonStatus
 {
     const STATE_OPEN = 'Open';

--- a/Domain/AmazonCaptureDetailsResponse.php
+++ b/Domain/AmazonCaptureDetailsResponse.php
@@ -15,6 +15,14 @@
  */
 namespace Amazon\Payment\Domain;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AmazonCaptureDetailsResponse extends AbstractAmazonCaptureResponse
 {
     private $resultKey = 'GetCaptureDetailsResult';

--- a/Domain/AmazonCaptureResponse.php
+++ b/Domain/AmazonCaptureResponse.php
@@ -15,6 +15,14 @@
  */
 namespace Amazon\Payment\Domain;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AmazonCaptureResponse extends AbstractAmazonCaptureResponse
 {
     private $resultKey = 'CaptureResult';

--- a/Domain/AmazonCaptureStatus.php
+++ b/Domain/AmazonCaptureStatus.php
@@ -15,6 +15,14 @@
  */
 namespace Amazon\Payment\Domain;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AmazonCaptureStatus extends AbstractAmazonStatus
 {
     const STATE_COMPLETED = 'Completed';

--- a/Domain/AmazonConstraint.php
+++ b/Domain/AmazonConstraint.php
@@ -15,6 +15,14 @@
  */
 namespace Amazon\Payment\Domain;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AmazonConstraint
 {
     const PAYMENT_METHOD_NOT_ALLOWED_ID = 'PaymentMethodNotAllowed';

--- a/Domain/AmazonGetOrderDetailsResponse.php
+++ b/Domain/AmazonGetOrderDetailsResponse.php
@@ -20,6 +20,14 @@ use Amazon\Payment\Domain\Details\AmazonOrderDetails;
 use Amazon\Payment\Domain\Details\AmazonOrderDetailsFactory;
 use AmazonPay\ResponseInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AmazonGetOrderDetailsResponse
 {
     /**

--- a/Domain/AmazonOrderStatus.php
+++ b/Domain/AmazonOrderStatus.php
@@ -15,6 +15,14 @@
  */
 namespace Amazon\Payment\Domain;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AmazonOrderStatus extends AbstractAmazonStatus
 {
     const STATE_DRAFT = 'Draft';

--- a/Domain/AmazonRefundDetailsResponse.php
+++ b/Domain/AmazonRefundDetailsResponse.php
@@ -15,6 +15,14 @@
  */
 namespace Amazon\Payment\Domain;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AmazonRefundDetailsResponse extends AbstractAmazonRefundResponse
 {
     private $resultKey = 'GetRefundDetailsResult';

--- a/Domain/AmazonRefundResponse.php
+++ b/Domain/AmazonRefundResponse.php
@@ -17,6 +17,14 @@ namespace Amazon\Payment\Domain;
 
 use AmazonPay\ResponseInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AmazonRefundResponse extends AbstractAmazonRefundResponse
 {
     private $resultKey = 'RefundResult';

--- a/Domain/AmazonRefundStatus.php
+++ b/Domain/AmazonRefundStatus.php
@@ -15,6 +15,14 @@
  */
 namespace Amazon\Payment\Domain;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AmazonRefundStatus extends AbstractAmazonStatus
 {
     const STATE_PENDING = 'Pending';

--- a/Domain/AmazonSetOrderDetailsResponse.php
+++ b/Domain/AmazonSetOrderDetailsResponse.php
@@ -18,6 +18,14 @@ namespace Amazon\Payment\Domain;
 use Amazon\Core\Exception\AmazonServiceUnavailableException;
 use AmazonPay\ResponseInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AmazonSetOrderDetailsResponse
 {
     /**

--- a/Domain/Details/AmazonAuthorizationDetails.php
+++ b/Domain/Details/AmazonAuthorizationDetails.php
@@ -18,6 +18,14 @@ namespace Amazon\Payment\Domain\Details;
 use Amazon\Payment\Domain\AmazonAuthorizationStatus;
 use Amazon\Payment\Domain\AmazonAuthorizationStatusFactory;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AmazonAuthorizationDetails
 {
     /**

--- a/Domain/Details/AmazonCaptureDetails.php
+++ b/Domain/Details/AmazonCaptureDetails.php
@@ -18,6 +18,14 @@ namespace Amazon\Payment\Domain\Details;
 use Amazon\Payment\Domain\AmazonCaptureStatus;
 use Amazon\Payment\Domain\AmazonCaptureStatusFactory;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AmazonCaptureDetails
 {
     /**

--- a/Domain/Details/AmazonOrderDetails.php
+++ b/Domain/Details/AmazonOrderDetails.php
@@ -18,6 +18,14 @@ namespace Amazon\Payment\Domain\Details;
 use Amazon\Payment\Domain\AmazonOrderStatus;
 use Amazon\Payment\Domain\AmazonOrderStatusFactory;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AmazonOrderDetails
 {
     /**

--- a/Domain/Details/AmazonRefundDetails.php
+++ b/Domain/Details/AmazonRefundDetails.php
@@ -18,6 +18,14 @@ namespace Amazon\Payment\Domain\Details;
 use Amazon\Payment\Domain\AmazonRefundStatus;
 use Amazon\Payment\Domain\AmazonRefundStatusFactory;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AmazonRefundDetails
 {
     /**

--- a/Domain/Validator/AmazonAuthorization.php
+++ b/Domain/Validator/AmazonAuthorization.php
@@ -22,6 +22,13 @@ use Amazon\Payment\Domain\Details\AmazonAuthorizationDetails;
 /**
  * Class AmazonAuthorization
  * validates Amazon Pay status during cron updates.
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class AmazonAuthorization
 {

--- a/Exception/AuthorizationExpiredException.php
+++ b/Exception/AuthorizationExpiredException.php
@@ -15,6 +15,14 @@
  */
 namespace Amazon\Payment\Exception;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AuthorizationExpiredException extends \Exception
 {
     public function __construct($message = '', $code = 0, \Exception $previous = null)

--- a/Exception/CapturePendingException.php
+++ b/Exception/CapturePendingException.php
@@ -15,6 +15,14 @@
  */
 namespace Amazon\Payment\Exception;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class CapturePendingException extends \Exception
 {
     public function __construct($message = '', $code = 0, \Exception $previous = null)

--- a/Exception/HardDeclineException.php
+++ b/Exception/HardDeclineException.php
@@ -18,6 +18,14 @@ namespace Amazon\Payment\Exception;
 use Magento\Framework\Exception\LocalizedException;
 use Magento\Framework\Phrase;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class HardDeclineException extends LocalizedException
 {
     public function __construct(Phrase $phrase, \Exception $cause = null)

--- a/Exception/SoftDeclineException.php
+++ b/Exception/SoftDeclineException.php
@@ -18,6 +18,14 @@ namespace Amazon\Payment\Exception;
 use Magento\Framework\Exception\LocalizedException;
 use Magento\Framework\Phrase;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class SoftDeclineException extends LocalizedException
 {
     public function __construct(Phrase $phrase, \Exception $cause = null)

--- a/Exception/TransactionTimeoutException.php
+++ b/Exception/TransactionTimeoutException.php
@@ -18,6 +18,14 @@ namespace Amazon\Payment\Exception;
 use Magento\Framework\Exception\LocalizedException;
 use Magento\Framework\Phrase;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class TransactionTimeoutException extends HardDeclineException
 {
     public function __construct(Phrase $phrase, \Exception $cause = null)

--- a/Gateway/Command/AmazonAuthCommand.php
+++ b/Gateway/Command/AmazonAuthCommand.php
@@ -27,6 +27,13 @@ use Amazon\Payment\Gateway\Config\Config;
  *
  * Enables customized error handling for Amazon Payment
  * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class AmazonAuthCommand implements CommandInterface
 {

--- a/Gateway/Command/CaptureStrategyCommand.php
+++ b/Gateway/Command/CaptureStrategyCommand.php
@@ -29,6 +29,14 @@ use Amazon\Core\Helper\Data;
 use Amazon\Core\Logger\ExceptionLogger;
 use Amazon\Payment\Gateway\Data\Order\OrderAdapterFactory;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class CaptureStrategyCommand implements CommandInterface
 {
 

--- a/Gateway/Config/Config.php
+++ b/Gateway/Config/Config.php
@@ -8,6 +8,13 @@ namespace Amazon\Payment\Gateway\Config;
 /**
  * Class Config
  * @package Amazon\Payment\Gateway\Config
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class Config extends \Magento\Payment\Gateway\Config\Config
 {

--- a/Gateway/Data/Order/OrderAdapter.php
+++ b/Gateway/Data/Order/OrderAdapter.php
@@ -25,6 +25,13 @@ use Amazon\Core\Helper\Data;
 
 /**
  * Class OrderAdapter
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class OrderAdapter implements OrderAdapterInterface
 {

--- a/Gateway/Helper/SubjectReader.php
+++ b/Gateway/Helper/SubjectReader.php
@@ -26,6 +26,13 @@ use Magento\Payment\Gateway\Data\PaymentDataObjectInterface;
 /**
  * Class SubjectReader
  * Consolidates commonly used calls
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class SubjectReader
 {

--- a/Gateway/Http/Client/AbstractClient.php
+++ b/Gateway/Http/Client/AbstractClient.php
@@ -28,6 +28,14 @@ use Amazon\Payment\Model\Adapter\AmazonPaymentAdapter;
  * Class AbstractClient
  * Base class for gateway client classes
  */
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 abstract class AbstractClient implements ClientInterface
 {
     /**

--- a/Gateway/Http/Client/AuthorizeClient.php
+++ b/Gateway/Http/Client/AuthorizeClient.php
@@ -19,6 +19,13 @@ namespace Amazon\Payment\Gateway\Http\Client;
 /**
  * Class Client
  * Amazon Pay authorization gateway client
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class AuthorizeClient extends AbstractClient
 {

--- a/Gateway/Http/Client/CaptureClient.php
+++ b/Gateway/Http/Client/CaptureClient.php
@@ -19,6 +19,13 @@ namespace Amazon\Payment\Gateway\Http\Client;
 /**
  * Class Client
  * Amazon Pay gateway capture client
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class CaptureClient extends AbstractClient
 {

--- a/Gateway/Http/Client/RefundClient.php
+++ b/Gateway/Http/Client/RefundClient.php
@@ -25,6 +25,13 @@ use Amazon\Payment\Domain\AmazonRefundResponseFactory;
 /**
  * Class RefundClient
  * Amazon Pay refund client
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class RefundClient implements ClientInterface
 {

--- a/Gateway/Http/Client/SettlementClient.php
+++ b/Gateway/Http/Client/SettlementClient.php
@@ -19,6 +19,13 @@ namespace Amazon\Payment\Gateway\Http\Client;
 /**
  * Class SettlementClient
  * Amazon Pay capture client
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class SettlementClient extends AbstractClient
 {

--- a/Gateway/Http/Client/VoidClient.php
+++ b/Gateway/Http/Client/VoidClient.php
@@ -19,6 +19,13 @@ namespace Amazon\Payment\Gateway\Http\Client;
 /**
  * Class VoidClient
  * Amazon Pay client for gateway cancel and void
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class VoidClient extends AbstractClient
 {

--- a/Gateway/Http/TransferFactory.php
+++ b/Gateway/Http/TransferFactory.php
@@ -19,6 +19,14 @@ use Magento\Payment\Gateway\Http\TransferBuilder;
 use Magento\Payment\Gateway\Http\TransferFactoryInterface;
 use Magento\Payment\Gateway\Http\TransferInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class TransferFactory implements TransferFactoryInterface
 {
     /**

--- a/Gateway/Request/AuthorizationRequestBuilder.php
+++ b/Gateway/Request/AuthorizationRequestBuilder.php
@@ -26,6 +26,14 @@ use Magento\Framework\DataObject;
 use Amazon\Payment\Plugin\AdditionalInformation;
 use Amazon\Core\Helper\CategoryExclusion;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AuthorizationRequestBuilder implements BuilderInterface
 {
     /**

--- a/Gateway/Request/RefundRequestBuilder.php
+++ b/Gateway/Request/RefundRequestBuilder.php
@@ -26,6 +26,13 @@ use Amazon\Payment\Gateway\Data\Order\OrderAdapterFactory;
 /**
  * Class RefundRequestBuilder
  * Builds refund request for Amazon Pay
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class RefundRequestBuilder implements BuilderInterface
 {

--- a/Gateway/Request/SettlementRequestBuilder.php
+++ b/Gateway/Request/SettlementRequestBuilder.php
@@ -24,6 +24,14 @@ use Magento\Payment\Model\Method\Logger;
 use Magento\Sales\Api\OrderRepositoryInterface;
 use Magento\Quote\Api\CartRepositoryInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class SettlementRequestBuilder implements BuilderInterface
 {
     /**

--- a/Gateway/Request/VoidRequestBuilder.php
+++ b/Gateway/Request/VoidRequestBuilder.php
@@ -22,6 +22,14 @@ use Amazon\Payment\Gateway\Helper\SubjectReader;
 use Amazon\Core\Helper\Data;
 use Magento\Sales\Api\OrderRepositoryInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class VoidRequestBuilder implements BuilderInterface
 {
 

--- a/Gateway/Response/CompleteAuthHandler.php
+++ b/Gateway/Response/CompleteAuthHandler.php
@@ -22,6 +22,14 @@ use Amazon\Payment\Gateway\Helper\SubjectReader;
 use Amazon\Core\Helper\Data;
 use Amazon\Payment\Api\Data\PendingAuthorizationInterfaceFactory;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class CompleteAuthHandler implements HandlerInterface
 {
 

--- a/Gateway/Response/CompleteSaleHandler.php
+++ b/Gateway/Response/CompleteSaleHandler.php
@@ -22,6 +22,14 @@ use Amazon\Payment\Gateway\Helper\SubjectReader;
 use Amazon\Core\Helper\Data;
 use Amazon\Payment\Api\Data\PendingAuthorizationInterfaceFactory;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class CompleteSaleHandler implements HandlerInterface
 {
 

--- a/Gateway/Response/RefundHandler.php
+++ b/Gateway/Response/RefundHandler.php
@@ -26,6 +26,13 @@ use Amazon\Payment\Api\Data\PendingRefundInterfaceFactory;
 /**
  * Class RefundHandler
  * Handles refund behavior for Amazon Pay
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class RefundHandler implements HandlerInterface
 {

--- a/Gateway/Response/SettlementHandler.php
+++ b/Gateway/Response/SettlementHandler.php
@@ -23,6 +23,14 @@ use Amazon\Payment\Gateway\Helper\SubjectReader;
 use Magento\Sales\Api\OrderRepositoryInterface;
 use Magento\Quote\Api\CartRepositoryInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class SettlementHandler implements HandlerInterface
 {
 

--- a/Gateway/Response/VoidHandler.php
+++ b/Gateway/Response/VoidHandler.php
@@ -22,6 +22,14 @@ use Magento\Payment\Model\Method\Logger;
 use Amazon\Payment\Gateway\Helper\SubjectReader;
 use Magento\Framework\Message\ManagerInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class VoidHandler implements HandlerInterface
 {
 

--- a/Gateway/Validator/AuthorizationValidator.php
+++ b/Gateway/Validator/AuthorizationValidator.php
@@ -23,6 +23,13 @@ use Amazon\Payment\Domain\AmazonConstraint;
 /**
  * Class AuthorizationValidator
  * Validates authorization calls during gateway payment
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class AuthorizationValidator extends AbstractValidator
 {

--- a/Gateway/Validator/ConstraintValidator.php
+++ b/Gateway/Validator/ConstraintValidator.php
@@ -21,6 +21,14 @@ use Magento\Payment\Gateway\Validator\ResultInterface;
 use Amazon\Payment\Gateway\Http\Client\Client;
 use Amazon\Payment\Domain\AmazonConstraint;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class ConstraintValidator extends AbstractValidator
 {
 

--- a/Helper/Address.php
+++ b/Helper/Address.php
@@ -23,6 +23,14 @@ use Magento\Customer\Api\Data\RegionInterfaceFactory;
 use Magento\Directory\Model\RegionFactory;
 use Magento\Framework\App\Config\ScopeConfigInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class Address
 {
     /**

--- a/Helper/Data.php
+++ b/Helper/Data.php
@@ -19,6 +19,14 @@ use Magento\Framework\App\Helper\AbstractHelper;
 use Magento\Framework\App\Helper\Context;
 use Magento\Framework\Module\ModuleListInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class Data extends AbstractHelper
 {
     const MODULE_CODE = 'Amazon_Payment';

--- a/Helper/Email.php
+++ b/Helper/Email.php
@@ -23,6 +23,14 @@ use Magento\Sales\Model\Order;
 use Magento\Framework\Mail\Template\TransportBuilderFactory;
 use Magento\Store\Model\ScopeInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class Email extends AbstractHelper
 {
     /**

--- a/Helper/Shortcut/CheckoutValidator.php
+++ b/Helper/Shortcut/CheckoutValidator.php
@@ -16,6 +16,14 @@
 
 namespace Amazon\Payment\Helper\Shortcut;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class CheckoutValidator implements ValidatorInterface
 {
     /**

--- a/Helper/Shortcut/Factory.php
+++ b/Helper/Shortcut/Factory.php
@@ -16,6 +16,14 @@
 
 namespace Amazon\Payment\Helper\Shortcut;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class Factory
 {
     /**

--- a/Helper/Shortcut/Validator.php
+++ b/Helper/Shortcut/Validator.php
@@ -18,6 +18,13 @@ namespace Amazon\Payment\Helper\Shortcut;
 
 /**
  * @SuppressWarnings(PHPMD.UnusedPrivateField)
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class Validator implements ValidatorInterface
 {

--- a/Ipn/IpnHandler.php
+++ b/Ipn/IpnHandler.php
@@ -17,6 +17,14 @@ namespace Amazon\Payment\Ipn;
 
 use AmazonPay\IpnHandler as AmazonIpnHandler;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class IpnHandler extends AmazonIpnHandler
 {
     /**

--- a/Ipn/IpnHandlerFactory.php
+++ b/Ipn/IpnHandlerFactory.php
@@ -21,6 +21,14 @@ use AmazonPay\IpnHandlerInterface;
 use Psr\Log\LoggerAwareInterface;
 use Psr\Log\LoggerInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class IpnHandlerFactory implements IpnHandlerFactoryInterface
 {
     /**

--- a/Model/Adapter/AmazonPaymentAdapter.php
+++ b/Model/Adapter/AmazonPaymentAdapter.php
@@ -32,6 +32,13 @@ use Magento\Framework\App\ObjectManager;
  * Class AmazonPaymentAdapter
  * Use \Amazon\Payment\Model\Adapter\AmazonPaymentAdapterFactory to create new instance of adapter.
  * @codeCoverageIgnore
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class AmazonPaymentAdapter
 {

--- a/Model/AddressManagement.php
+++ b/Model/AddressManagement.php
@@ -37,6 +37,13 @@ use Psr\Log\LoggerInterface;
 
 /**
  * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class AddressManagement implements AddressManagementInterface
 {

--- a/Model/Ipn/AuthorizationProcessor.php
+++ b/Model/Ipn/AuthorizationProcessor.php
@@ -22,6 +22,14 @@ use Amazon\Payment\Domain\Details\AmazonAuthorizationDetailsFactory;
 use Amazon\Payment\Model\ResourceModel\PendingAuthorization\CollectionFactory;
 use Magento\Store\Model\StoreManagerInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AuthorizationProcessor implements ProcessorInterface
 {
     /**

--- a/Model/Ipn/CaptureProcessor.php
+++ b/Model/Ipn/CaptureProcessor.php
@@ -22,6 +22,14 @@ use Amazon\Payment\Domain\Details\AmazonCaptureDetailsFactory;
 use Amazon\Payment\Model\ResourceModel\PendingCapture\CollectionFactory;
 use Magento\Store\Model\StoreManagerInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class CaptureProcessor implements ProcessorInterface
 {
     /**

--- a/Model/Ipn/CompositeProcessor.php
+++ b/Model/Ipn/CompositeProcessor.php
@@ -18,6 +18,14 @@ namespace Amazon\Payment\Model\Ipn;
 use Amazon\Payment\Api\Ipn\CompositeProcessorInterface;
 use Amazon\Payment\Api\Ipn\ProcessorInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class CompositeProcessor implements CompositeProcessorInterface
 {
     /**

--- a/Model/Ipn/OrderProcessor.php
+++ b/Model/Ipn/OrderProcessor.php
@@ -23,6 +23,14 @@ use Amazon\Payment\Model\ResourceModel\OrderLink;
 use Amazon\Payment\Model\ResourceModel\PendingAuthorization\CollectionFactory;
 use Magento\Store\Model\StoreManagerInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class OrderProcessor implements ProcessorInterface
 {
     /**

--- a/Model/Ipn/RefundProcessor.php
+++ b/Model/Ipn/RefundProcessor.php
@@ -22,6 +22,14 @@ use Amazon\Payment\Model\QueuedRefundUpdater;
 use Amazon\Payment\Model\ResourceModel\PendingRefund\CollectionFactory;
 use Magento\Store\Model\StoreManagerInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class RefundProcessor implements ProcessorInterface
 {
     /**

--- a/Model/Method/AmazonLoginMethod.php
+++ b/Model/Method/AmazonLoginMethod.php
@@ -20,6 +20,13 @@ use Amazon\Core\Helper\Data as AmazonCoreHelper;
 /**
  * Class AmazonLoginMethod
  * For display of Amazon Pay in payment method list
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class AmazonLoginMethod extends \Magento\Payment\Model\Method\AbstractMethod
 {

--- a/Model/OrderInformationManagement.php
+++ b/Model/OrderInformationManagement.php
@@ -37,6 +37,13 @@ use Magento\Framework\App\ObjectManager;
 
 /**
  * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class OrderInformationManagement implements OrderInformationManagementInterface
 {
@@ -259,7 +266,7 @@ class OrderInformationManagement implements OrderInformationManagementInterface
     public function removeOrderReference()
     {
         $quote = $this->session->getQuote();
-        
+
         if ($quote->getId()) {
             $quoteLink = $this->quoteLinkFactory->create()->load($quote->getId(), 'quote_id');
 

--- a/Model/OrderLink.php
+++ b/Model/OrderLink.php
@@ -19,6 +19,14 @@ use Amazon\Payment\Api\Data\OrderLinkInterface;
 use Amazon\Payment\Model\ResourceModel\OrderLink as OrderLinkResourceModel;
 use Magento\Framework\Model\AbstractModel;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class OrderLink extends AbstractModel implements OrderLinkInterface
 {
     protected function _construct()

--- a/Model/PaymentManagement.php
+++ b/Model/PaymentManagement.php
@@ -32,6 +32,13 @@ use Magento\Sales\Api\TransactionRepositoryInterface;
 
 /**
  * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class PaymentManagement
 {

--- a/Model/PaymentManagement/AbstractOperation.php
+++ b/Model/PaymentManagement/AbstractOperation.php
@@ -34,6 +34,14 @@ use Magento\Sales\Model\Order\Payment\Transaction;
 use Magento\Store\Model\ScopeInterface;
 use Magento\Store\Model\StoreManagerInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 abstract class AbstractOperation
 {
     /**

--- a/Model/PaymentManagement/Authorization.php
+++ b/Model/PaymentManagement/Authorization.php
@@ -46,6 +46,13 @@ use Amazon\Payment\Exception\TransactionTimeoutException;
 
 /**
  * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class Authorization extends AbstractOperation
 {

--- a/Model/PaymentManagement/Capture.php
+++ b/Model/PaymentManagement/Capture.php
@@ -40,6 +40,13 @@ use Psr\Log\LoggerInterface;
 
 /**
  * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class Capture extends AbstractOperation
 {

--- a/Model/PendingAuthorization.php
+++ b/Model/PendingAuthorization.php
@@ -25,6 +25,14 @@ use Magento\Framework\Registry;
 use Magento\Framework\Stdlib\DateTime\DateTimeFactory;
 use Magento\Sales\Api\Data\OrderInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class PendingAuthorization extends AbstractModel implements PendingAuthorizationInterface
 {
     /**

--- a/Model/PendingCapture.php
+++ b/Model/PendingCapture.php
@@ -24,6 +24,14 @@ use Magento\Framework\Model\ResourceModel\AbstractResource;
 use Magento\Framework\Registry;
 use Magento\Framework\Stdlib\DateTime\DateTimeFactory;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class PendingCapture extends AbstractModel implements PendingCaptureInterface
 {
     /**

--- a/Model/PendingRefund.php
+++ b/Model/PendingRefund.php
@@ -24,6 +24,14 @@ use Magento\Framework\Model\ResourceModel\AbstractResource;
 use Magento\Framework\Registry;
 use Magento\Framework\Stdlib\DateTime\DateTimeFactory;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class PendingRefund extends AbstractModel implements PendingRefundInterface
 {
     /**

--- a/Model/QueuedRefundUpdater.php
+++ b/Model/QueuedRefundUpdater.php
@@ -27,6 +27,14 @@ use Magento\Sales\Api\OrderRepositoryInterface;
 use Magento\Store\Model\StoreManagerInterface;
 use Psr\Log\LoggerInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class QueuedRefundUpdater
 {
     /**

--- a/Model/QuoteLink.php
+++ b/Model/QuoteLink.php
@@ -19,6 +19,14 @@ use Amazon\Payment\Api\Data\QuoteLinkInterface;
 use Amazon\Payment\Model\ResourceModel\QuoteLink as QuoteLinkResourceModel;
 use Magento\Framework\Model\AbstractModel;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class QuoteLink extends AbstractModel implements QuoteLinkInterface
 {
     protected function _construct()

--- a/Model/QuoteLinkManagement.php
+++ b/Model/QuoteLinkManagement.php
@@ -20,6 +20,14 @@ use Amazon\Payment\Api\Data\QuoteLinkInterfaceFactory;
 use Magento\Quote\Api\Data\CartExtensionFactory;
 use Magento\Quote\Api\Data\CartInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class QuoteLinkManagement implements QuoteLinkManagementInterface
 {
     /**

--- a/Model/ResourceModel/OrderLink.php
+++ b/Model/ResourceModel/OrderLink.php
@@ -17,6 +17,14 @@ namespace Amazon\Payment\Model\ResourceModel;
 
 use Magento\Framework\Model\ResourceModel\Db\AbstractDb;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class OrderLink extends AbstractDb
 {
     const TABLE_NAME = 'amazon_sales_order';

--- a/Model/ResourceModel/PendingAuthorization.php
+++ b/Model/ResourceModel/PendingAuthorization.php
@@ -18,6 +18,14 @@ namespace Amazon\Payment\Model\ResourceModel;
 use Amazon\Payment\Api\Data\PendingAuthorizationInterface;
 use Magento\Framework\Model\ResourceModel\Db\AbstractDb;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class PendingAuthorization extends AbstractDb
 {
     const TABLE_NAME = 'amazon_pending_authorization';

--- a/Model/ResourceModel/PendingAuthorization/Collection.php
+++ b/Model/ResourceModel/PendingAuthorization/Collection.php
@@ -22,6 +22,14 @@ use Magento\Framework\Model\ResourceModel\Db\Collection\AbstractCollection;
 use Amazon\Payment\Model\PendingAuthorization as PendingAuthorizationModel;
 use Amazon\Payment\Model\ResourceModel\PendingAuthorization as PendingAuthorizationResourceModel;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class Collection extends AbstractCollection
 {
     protected function _construct()

--- a/Model/ResourceModel/PendingCapture.php
+++ b/Model/ResourceModel/PendingCapture.php
@@ -18,6 +18,14 @@ namespace Amazon\Payment\Model\ResourceModel;
 use Amazon\Payment\Api\Data\PendingCaptureInterface;
 use Magento\Framework\Model\ResourceModel\Db\AbstractDb;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class PendingCapture extends AbstractDb
 {
     const TABLE_NAME = 'amazon_pending_capture';

--- a/Model/ResourceModel/PendingCapture/Collection.php
+++ b/Model/ResourceModel/PendingCapture/Collection.php
@@ -21,6 +21,14 @@ use Magento\Framework\Model\ResourceModel\Db\Collection\AbstractCollection;
 use Amazon\Payment\Model\PendingCapture as PendingCaptureModel;
 use Amazon\Payment\Model\ResourceModel\PendingCapture as PendingCaptureResourceModel;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class Collection extends AbstractCollection
 {
     protected function _construct()

--- a/Model/ResourceModel/PendingRefund.php
+++ b/Model/ResourceModel/PendingRefund.php
@@ -19,6 +19,14 @@ use Amazon\Payment\Api\Data\PendingRefundInterface;
 use Amazon\Payment\Setup\UpgradeSchema as AmazonSchemaHelper;
 use Magento\Framework\Model\ResourceModel\Db\AbstractDb;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class PendingRefund extends AbstractDb
 {
     const TABLE_NAME = AmazonSchemaHelper::PENDING_REFUND_TABLE_NAME;

--- a/Model/ResourceModel/PendingRefund/Collection.php
+++ b/Model/ResourceModel/PendingRefund/Collection.php
@@ -20,6 +20,14 @@ use Amazon\Payment\Model\PendingRefund as PendingRefundModel;
 use Amazon\Payment\Model\ResourceModel\PendingRefund as PendingRefundResourceModel;
 use Magento\Framework\Model\ResourceModel\Db\Collection\AbstractCollection;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class Collection extends AbstractCollection
 {
 

--- a/Model/ResourceModel/QuoteLink.php
+++ b/Model/ResourceModel/QuoteLink.php
@@ -17,6 +17,14 @@ namespace Amazon\Payment\Model\ResourceModel;
 
 use Magento\Framework\Model\ResourceModel\Db\AbstractDb;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class QuoteLink extends AbstractDb
 {
     const TABLE_NAME = 'amazon_quote';

--- a/Observer/AddAmazonButton.php
+++ b/Observer/AddAmazonButton.php
@@ -20,6 +20,14 @@ use Magento\Framework\Event\Observer;
 use Amazon\Core\Helper\Data;
 use Amazon\Payment\Helper\Shortcut\Factory as ShortcutFactory;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AddAmazonButton implements ObserverInterface
 {
     /**

--- a/Observer/DataAssignObserver.php
+++ b/Observer/DataAssignObserver.php
@@ -20,6 +20,14 @@ use Magento\Payment\Observer\AbstractDataAssignObserver;
 use Magento\Quote\Api\Data\PaymentInterface;
 use Amazon\Payment\Plugin\AdditionalInformation;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class DataAssignObserver extends AbstractDataAssignObserver
 {
 

--- a/Observer/HardDeclinedEmailSender.php
+++ b/Observer/HardDeclinedEmailSender.php
@@ -19,6 +19,14 @@ use Amazon\Payment\Helper\Email;
 use Magento\Framework\Event\Observer;
 use Magento\Framework\Event\ObserverInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class HardDeclinedEmailSender implements ObserverInterface
 {
     /**

--- a/Observer/IgnoreBillingAddressValidation.php
+++ b/Observer/IgnoreBillingAddressValidation.php
@@ -19,6 +19,14 @@ use Amazon\Payment\Gateway\Config\Config as AmazonPayment;
 use Magento\Framework\Event\Observer;
 use Magento\Framework\Event\ObserverInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class IgnoreBillingAddressValidation implements ObserverInterface
 {
     public function execute(Observer $observer)

--- a/Observer/KlarnaKcoOverride.php
+++ b/Observer/KlarnaKcoOverride.php
@@ -20,6 +20,14 @@ use Magento\Framework\Event\Observer;
 use Amazon\Core\Helper\Data;
 use Amazon\Login\Helper\Session;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class KlarnaKcoOverride implements ObserverInterface
 {
     /**

--- a/Observer/LoadOrder.php
+++ b/Observer/LoadOrder.php
@@ -24,6 +24,14 @@ use Magento\Sales\Api\Data\OrderInterface;
 use Amazon\Payment\Api\Data\QuoteLinkInterfaceFactory;
 use Amazon\Payment\Model\Adapter\AmazonPaymentAdapter;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class LoadOrder implements ObserverInterface
 {
     /**

--- a/Observer/OrderPaymentCapture.php
+++ b/Observer/OrderPaymentCapture.php
@@ -22,6 +22,13 @@ use Magento\Framework\Event\ObserverInterface;
 /**
  * Class OrderPaymentCapture
  * @package Amazon\Payment\Observer
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class OrderPaymentCapture implements ObserverInterface
 {

--- a/Observer/SandboxSimulation.php
+++ b/Observer/SandboxSimulation.php
@@ -21,6 +21,14 @@ use Magento\Framework\Event\Observer;
 use Magento\Framework\Event\ObserverInterface;
 use Magento\Payment\Model\InfoInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class SandboxSimulation implements ObserverInterface
 {
     /**

--- a/Observer/SoftDeclinedEmailSender.php
+++ b/Observer/SoftDeclinedEmailSender.php
@@ -19,6 +19,14 @@ use Amazon\Payment\Helper\Email;
 use Magento\Framework\Event\Observer;
 use Magento\Framework\Event\ObserverInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class SoftDeclinedEmailSender implements ObserverInterface
 {
     /**

--- a/Plugin/AdditionalInformation.php
+++ b/Plugin/AdditionalInformation.php
@@ -17,6 +17,14 @@ namespace Amazon\Payment\Plugin;
 
 use Magento\Quote\Model\Quote\Payment;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class AdditionalInformation
 {
     const KEY_SANDBOX_SIMULATION_REFERENCE = 'sandbox_simulation_reference';

--- a/Plugin/CheckoutProcessor.php
+++ b/Plugin/CheckoutProcessor.php
@@ -18,6 +18,14 @@ namespace Amazon\Payment\Plugin;
 use Amazon\Core\Helper\Data;
 use Magento\Checkout\Model\Session;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class CheckoutProcessor
 {
     /**

--- a/Plugin/ConfirmOrderReference.php
+++ b/Plugin/ConfirmOrderReference.php
@@ -34,6 +34,13 @@ use Magento\Quote\Api\CartRepositoryInterface;
  * Class ConfirmOrderReference
  *
  * Confirm the OrderReference when payment details are saved
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class ConfirmOrderReference
 {

--- a/Plugin/DisableAmazonPaymentMethod.php
+++ b/Plugin/DisableAmazonPaymentMethod.php
@@ -13,6 +13,13 @@ use Magento\Checkout\Model\Session;
 /**
  * Class DisableAmazonPaymentMethod
  * Plugin removes Amazon Payment Method if cart contains only virtual products.
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class DisableAmazonPaymentMethod
 {

--- a/Plugin/OrderCurrencyComment.php
+++ b/Plugin/OrderCurrencyComment.php
@@ -22,6 +22,13 @@ use Amazon\Payment\Gateway\Config\Config;
 /**
  * Class OrderCurrencyComment
  * @package Amazon\Payment\Plugin
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class OrderCurrencyComment
 {

--- a/Plugin/QuoteRepository.php
+++ b/Plugin/QuoteRepository.php
@@ -20,6 +20,14 @@ use Amazon\Payment\Api\QuoteLinkManagementInterface;
 use Magento\Quote\Api\CartRepositoryInterface;
 use Magento\Quote\Api\Data\CartInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class QuoteRepository
 {
     /**

--- a/Plugin/ShippingInformationManagement.php
+++ b/Plugin/ShippingInformationManagement.php
@@ -23,6 +23,14 @@ use Magento\Checkout\Api\ShippingInformationManagementInterface;
 use Magento\Quote\Api\CartRepositoryInterface;
 use Amazon\Login\Helper\Session as LoginSessionHelper;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class ShippingInformationManagement
 {
     /**

--- a/Plugin/WebapiErrorProcessor.php
+++ b/Plugin/WebapiErrorProcessor.php
@@ -18,6 +18,14 @@ namespace Amazon\Payment\Plugin;
 use Amazon\Core\Exception\AmazonWebapiException;
 use Magento\Framework\Webapi\ErrorProcessor;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class WebapiErrorProcessor
 {
     /**

--- a/Setup/UpgradeData.php
+++ b/Setup/UpgradeData.php
@@ -21,6 +21,14 @@ use Magento\Framework\Setup\ModuleContextInterface;
 use Magento\Framework\Setup\ModuleDataSetupInterface;
 use Magento\Framework\Setup\UpgradeDataInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class UpgradeData implements UpgradeDataInterface
 {
     /**

--- a/Setup/UpgradeSchema.php
+++ b/Setup/UpgradeSchema.php
@@ -32,6 +32,14 @@ use Magento\Framework\Setup\SchemaSetupInterface;
 use Magento\Framework\Setup\UpgradeSchemaInterface;
 use Magento\Sales\Api\Data\TransactionInterface;
 
+/**
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
+ */
 class UpgradeSchema implements UpgradeSchemaInterface
 {
 

--- a/Test/Unit/Gateway/Command/CaptureStrategyCommandTest.php
+++ b/Test/Unit/Gateway/Command/CaptureStrategyCommandTest.php
@@ -35,6 +35,13 @@ use PHPUnit_Framework_MockObject_MockObject as MockObject;
  * Class CaptureStrategyCommandTest
  *
  * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class CaptureStrategyCommandTest extends \PHPUnit\Framework\TestCase
 {

--- a/Test/Unit/Observer/AddAmazonButtonTest.php
+++ b/Test/Unit/Observer/AddAmazonButtonTest.php
@@ -26,6 +26,13 @@ use Amazon\Payment\Observer\AddAmazonButton;
  * Class AddAmazonButtonTest
  *
  * @see \Amazon\Payment\Observer\AddAmazonButton
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class AddAmazonButtonTest extends \PHPUnit\Framework\TestCase
 {

--- a/Test/Unit/Observer/DataAssignObserverTest.php
+++ b/Test/Unit/Observer/DataAssignObserverTest.php
@@ -24,6 +24,13 @@ use Magento\Quote\Api\Data\PaymentInterface;
 
 /**
  * Class DataAssignObserverTest
+ *
+ * @deprecated As of February 2021, this Legacy Amazon Pay plugin has been
+ * deprecated, in favor of a newer Amazon Pay version available through GitHub
+ * and Magento Marketplace. Please download the new plugin for automatic
+ * updates and to continue providing your customers with a seamless checkout
+ * experience. Please see https://pay.amazon.com/help/E32AAQBC2FY42HS for details
+ * and installation instructions.
  */
 class DataAssignObserverTest extends \PHPUnit\Framework\TestCase
 {

--- a/composer.json
+++ b/composer.json
@@ -2,14 +2,15 @@
   "name": "amzn/amazon-pay-module",
   "description": "Amazon Pay module",
   "type": "magento2-module",
-  "version": "3.7.4",
+  "version": "3.8.0",
+  "abandoned": "amzn/amazon-pay-v2-magento-2-module",
   "license": [
     "Apache-2.0"
   ],
   "require": {
     "php": "~7.1.3||~7.2.0||~7.3.0",
-    "amzn/amazon-pay-and-login-with-amazon-core-module": "^3.7.4",
-    "amzn/login-with-amazon-module": "^3.7.4",
+    "amzn/amazon-pay-and-login-with-amazon-core-module": "^3.8.0",
+    "amzn/login-with-amazon-module": "^3.8.0",
     "magento/framework": "^102",
     "magento/module-customer": "^102",
     "magento/module-eav": "^102",

--- a/etc/module.xml
+++ b/etc/module.xml
@@ -16,7 +16,7 @@
  */
 -->
 <config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Module/etc/module.xsd">
-    <module name="Amazon_Payment" setup_version="3.7.4" >
+    <module name="Amazon_Payment" setup_version="3.8.0" >
         <sequence>
             <module name="Amazon_Core"/>
             <module name="Amazon_Login"/>

